### PR TITLE
Classes for the CracFile refactoring

### PR DESCRIPTION
**Pull request by terriervik**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Refactoring of the Crac file

### CODE REVIEW

File: data/crac-file/crac-impl/pom.xml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        <version>1.3.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>farao-crac-impl</artifactId>
+
+    <dependencies>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>crac-api</artifactId>
```
Comment (TheMaskedTurtle) [inline data/crac-file/crac-impl/pom.xml]: `farao-crac-api` for consistency with module name

File: data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Side.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * @author Viktor Terrier {@literal <viktor.terrier at rte-france.com>}
+ */
+
+public enum Side {
+    LEFT,
+    RIGHT,
+    DOWN,
+    MIDDLE,
+    UP
```
Comment (TheMaskedTurtle) [inline data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Side.java]: You can remove that

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/network_action/NetworkAction.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        this.networkActions = networkActions;
+    }
+
+    public List<ApplicableNetworkAction> getNetworkActions() {
+        return networkActions;
+    }
+
+    @Override
+    public void apply(Network network) {
+        throw new UnsupportedOperationException();
```
Comment (TheMaskedTurtle) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/network_action/NetworkAction.java]: You can implement this one by calling apply() method on all the elements of the networkActions list

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_action/RangeAction.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Override
+    public double getMaxValue(Network network) {
+        return 0;
+    }
+
+    @Override
+    public void apply(Network network, double setpoint) {
+        throw new UnsupportedOperationException();
```
Comment (TheMaskedTurtle) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_action/RangeAction.java]: You can implement this one by calling apply() method on all the elements of the elementaryRangeActions list

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_action/RangeAction.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * Group of simultaneously applied range remedial actions.
+ *
+ * @author Viktor Terrier {@literal <viktor.terrier at rte-france.com>}
+ */
+public class RangeAction extends AbstractRemedialAction implements Range, ApplicableRangeAction {
+
+    private List<Range> ranges;
+    private List<ApplicableRangeAction> elementaryRangeActions;
```
Comment (TheMaskedTurtle) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_action/RangeAction.java]: rename applicableRangeActions for consistency ?

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/network_action/NetworkAction.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.List;
+
+/**
+ * Group of simple elementary remedial actions (setpoint, open/close, ...).
+ *
+ * @author Viktor Terrier {@literal <viktor.terrier at rte-france.com>}
+ */
+public class NetworkAction implements ApplicableNetworkAction {
+
+    private List<ApplicableNetworkAction> networkActions;
```
Comment (TheMaskedTurtle) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/network_action/NetworkAction.java]: Rename applicableNetworkActions for consistency ?

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_domain/Range.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * Range.
+ *
+ * @author Viktor Terrier {@literal <viktor.terrier at rte-france.com>}
+ */
+public interface Range {
+
+    public double getMinValue(Network network);
+
+    public double getMaxValue(Network network);
```
Comment (TheMaskedTurtle) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_domain/Range.java]: Public modifiers are redundant, so you can remove them. Same comment for every interface.

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/threshold/AbstractThreshold.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package com.farao_community.farao.data.crac_impl.remedial_action.threshold;
```
Comment (TheMaskedTurtle) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/threshold/AbstractThreshold.java]: It might more consistent to move threshold package out of remedial_action package because thresholds can be related to Cnecs as well.

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/network_action/NetworkAction.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+import com.powsybl.iidm.network.Network;
+import java.util.List;
+
+/**
+ * Group of simple elementary remedial actions (setpoint, open/close, ...).
+ *
+ * @author Viktor Terrier {@literal <viktor.terrier at rte-france.com>}
+ */
+public class NetworkAction implements ApplicableNetworkAction {
```
Comment (TheMaskedTurtle) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/network_action/NetworkAction.java]: NetworkAction should extends AbstractRemedialAction

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/usage_rule/OnConstraint.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * The remedial action is available only after a specific constraint, with a given method
+ * and in a given state.
+ *
+ * @author Viktor Terrier {@literal <viktor.terrier at rte-france.com>}
+ */
+public final class OnConstraint extends AbstractUsageRule {
+
+    private Cnec constraint;
```
Comment (TheMaskedTurtle) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/usage_rule/OnConstraint.java]: Rename cnec ?

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_action/Countertrading.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+package com.farao_community.farao.data.crac_impl.remedial_action.range_action;
+
+import com.powsybl.iidm.network.Network;
+
+/**
+ * Countertrading remedial action.
+ *
+ * @author Xxx Xxx {@literal <xxx.xxx at rte-france.com>}
+ */
+
```
Comment (alexandrem-rte) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_action/Countertrading.java]: You can add your mail adress here ;)

File: data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Identifiable.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+ */
+package com.farao_community.farao.data.crac_api;
+
+import com.powsybl.commons.extensions.Extendable;
+
+/**
+ * An object that is part of the network model and that is identified uniquely
+ * by a <code>String</code> id.
+ *
+ * @author Viktor Terrier <viktor.terrier at rte-france.com>
```
Comment (murgeyseb) [inline data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Identifiable.java]: wrap <...> with {@literal}

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/AbstractIdentifiable.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.farao_community.farao.data.crac_impl;
+
+import com.powsybl.commons.extensions.AbstractExtendable;
+import com.farao_community.farao.data.crac_api.Identifiable;
+
+/**
+ *
+ * @author Viktor Terrier <viktor.terrier at rte-france.com>
```
Comment (murgeyseb) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/AbstractIdentifiable.java]: same remark, wrap <...> with {@literal}

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_domain/RelativeDynamicRange.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package com.farao_community.farao.data.crac_impl.remedial_action.range_domain;
+
+import com.powsybl.iidm.network.Network;
+
+/**
+ * Absolute validity domain.
```
Comment (murgeyseb) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_domain/RelativeDynamicRange.java]: Comments does not seem to describe this object

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_domain/RelativeFixedRange.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package com.farao_community.farao.data.crac_impl.remedial_action.range_domain;
+
+import com.powsybl.iidm.network.Network;
+
+/**
+ * Absolute validity domain.
```
Comment (murgeyseb) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_domain/RelativeFixedRange.java]: same remark than before

File: data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Identifiable.java [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    /**
+     * Get an the (optional) name  of the object.
+     */
+    String getName();
+}
+
+
+
+
+
```
Comment (murgeyseb) [inline data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Identifiable.java]: No need for so many empty lines !

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/NetworkElement.java [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package com.farao_community.farao.data.crac_impl;
+
+/**
+ * Element of the network in the CRAC file.
+ *
+ * @author Viktor Terruer {@literal <viktor.terrier at rte-france.com>}
```
Comment (alexandrem-rte) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/NetworkElement.java]: Shouldn't we add the following attributes to this class: list of AbstractRemedialAction, list of Cnec, list of Contingency ?

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/State.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+package com.farao_community.farao.data.crac_impl;
+
+import java.util.Optional;
+
+/**
+ * Business object for a state (instant and contingency) in the CRAC file.
+ *
+ * @author Viktor Terrier {@literal <viktor.terrier at rte-france.com>}
+ */
+public class State {
```
Comment (alexandrem-rte) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/State.java]: Shouldn't we add an attribute: list of AbstractUsageRule?

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/NetworkElement.java (hunk 17 above)
Comment (terriervik) [reply data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/NetworkElement.java]: We'll do it later if needed.

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/State.java (hunk 18 above)
Comment (terriervik) [reply data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/State.java]: We'll do it later if needed.

File: data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Direction.java [hunk 19]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package com.farao_community.farao.data.crac_api;
```
Comment (TheMaskedTurtle) [inline data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Direction.java]: Extract more classes to crac_api

File: data/crac-file/crac-api/src/test/java/com/farao_community/farao/data/crac_api/NetworkElementTest.java [hunk 20]
```diff
@@ -0,0 +1,17 @@
+package com.farao_community.farao.data.crac_api;
```
Comment (murgeyseb) [inline data/crac-file/crac-api/src/test/java/com/farao_community/farao/data/crac_api/NetworkElementTest.java]: Copyright header

File: data/crac-file/crac-api/src/test/java/com/farao_community/farao/data/crac_api/NetworkElementTest.java [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+public class NetworkElementTest {
```
Comment (murgeyseb) [inline data/crac-file/crac-api/src/test/java/com/farao_community/farao/data/crac_api/NetworkElementTest.java]: Author

File: data/crac-file/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/SimpleStateTest.java [hunk 22]
```diff
@@ -0,0 +1,44 @@
+package com.farao_community.farao.data.crac_impl;
```
Comment (murgeyseb) [inline data/crac-file/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/SimpleStateTest.java]: Copyright

File: data/crac-file/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/SimpleStateTest.java [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Optional;
+
+import static org.junit.Assert.assertEquals;
+
+public class SimpleStateTest {
```
Comment (murgeyseb) [inline data/crac-file/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/SimpleStateTest.java]: Author

File: data/crac-file/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/CracFileTest.java [hunk 24]
```diff
@@ -168,17 +167,17 @@ public void testCrac() {
...
 
-        Crac crac = new Crac("idCrac", "name", new ArrayList<>(), new ArrayList<>(), new ArrayList<>());
+        SimpleCrac crac = new SimpleCrac("idCrac", "name", new ArrayList<>(), new ArrayList<>(), new ArrayList<>());
 
         crac.setCnecs(cnecs);
         crac.addCnec(cnec2);
-        crac.setNetworkActions(new ArrayList<>(Arrays.asList(networkAction1)));
-        crac.addNetworkRemedialAction(networkAction2);
+        /*crac.setNetworkActions(new ArrayList<>(Arrays.asList(networkAction1)));
+        crac.addNetworkRemedialAction(networkAction2);*/
```
Comment (terriervik) [inline data/crac-file/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/CracFileTest.java]: This should be adapted or deleted.

File: data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/State.java [hunk 25]
```diff
@@ -0,0 +1,14 @@
+package com.farao_community.farao.data.crac_api;
```
Comment (terriervik) [inline data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/State.java]: Why did you delete the header and author?
